### ISSUE TITLE
Recaptcha should support strict CSP style-src rules.

### ISSUE THREAD

**Issue by hmosvoll** [google/recaptcha#107]
[The documentation](https://developers.google.com/recaptcha/docs/faq#im-using-content-security-policy-csp-on-my-website-how-can-i-configure-it-to-work-with-recaptcha) now advice developers to enable style-src='unsafe-inline'. 

Should not Recaptcha give an alternative that does not require developers to lower their sites security?

**Comment by SaltOfTheFlame** [comment]
is there any chance that this issue wil be solved someday?

**Comment by Sora2455** [comment]
The reCaptcha FAQ claims that adding a noonce to the initial script tag will allow reCaptcha to work with CSP:[ link here](https://developers.google.com/recaptcha/docs/faq#im-using-content-security-policy-csp-on-my-website-how-can-i-configure-it-to-work-with-recaptcha).
However, my own testing has shown that while the external scripts are loaded correctly, the inline styles are still broken, and thus the widget becomes unusable.

**Comment by Sora2455** [comment]
After doing some research, it appears that inline styles _are_ valid in a strict CSP policy, but only if they were set by JavaScript. [Link.](http://www.cspplayground.com/compliant_examples#style)

So while `<div id='example-inline-style' style='color:blue;></div>` is invalid under a strict CSP policy,
`<div id='example-inline-style' style='color:blue;></div>
<script noonce='noonce'>document.getElementById('example-inline-style').style = 'color:blue;';</script>`
achieves the same effect and is perfectly valid.
Could the reCaptcha styles be set this way?

**Comment by JasonThomasData** [comment]
FYI, this has not worked for me. I'm using ASP.NET, and using NWebsec to create a nonce. I can confirm that once my page loads, the nonce attribute has been added to the DOM element and the response headers contain that same nonce, but that rule does not apply to the style sheets that the api script tries to load.

The unsafe-inline directive works, but my employer would never approve that. Will report back if I find a solution.

**Comment by bradleyhodges** [comment]
Nudging this issue. It sort of bums out Google's whole stance on security. I refuse to change my Content Security Policy - impacting my site's security, simply because Google refuses to get with the times.

**Comment by IanOliver** [comment]
@bradleyhodges Same here. I'm not allowing inline styles just for recaptcha. Last commit was in March (nine months ago at time of writing), so development seems to have all but died?

**Comment by bradleyhodges** [comment]
Absolutely.. project is definitely dead, but I’m pretty sure that this repo is for the PHP library alone; meaning that we’d have to find another way to get Google’s attention.

It’s still a massive disappointment to see that Google isn’t maintaining security across the entire project. There isn’t even a way to specify a nonce or CSP compliance policy with their Javascript library, and that documentation is on Google’s own website.

**Comment by Avamander** [comment]
Any updates to this?

**Comment by rowan-m** [comment]
I'll try setting a CSP on the demo and see what happens.

**Comment by rowan-m** [comment]
I've been playing around locally and it would be useful to hear some feedback on what people are looking for here. You can take a quick look at the [`csp-example` branch](https://github.com/google/recaptcha/blob/csp-example/examples/recaptcha-v3-request-scores.php#L31-L40) where I'm doing this:
```php
$recaptchaNonce = base64_encode(openssl_random_pseudo_bytes(16));
$inlineNonce = base64_encode(openssl_random_pseudo_bytes(16));
$gaIncNonce = base64_encode(openssl_random_pseudo_bytes(16));
$gaCfgNonce = base64_encode(openssl_random_pseudo_bytes(16));

header("Content-Security-Policy: "
    ."script-src 'nonce-".$inlineNonce."' 'nonce-".$recaptchaNonce."' 'nonce-".$gaIncNonce."' 'nonce-".$gaCfgNonce."' 'strict-dynamic'; "
    ."object-src 'none'; "
    ."base-uri 'none'; "
);
```
To fire off a header:
```
Content-Security-Policy: script-src 'nonce-XGgRBiS8q/NL9/hMQaTUtQ==' 'nonce-6Hei0z/rbjQEvwHgCc7wXg==' 'nonce-5FQCJydN0DcsTkrGzCFU8g==' 'nonce-HS3M7LHHGSK6b5gxt01gmg==' 'strict-dynamic'; object-src 'none'; base-uri 'none';
```
Including the library with
```php
<script src="https://www.google.com/recaptcha/api.js?render=<?php echo $siteKey; ?>" nonce="<?php echo $recaptchaNonce; ?>"></script>
```
Generating
```
<script src="https://www.google.com/recaptcha/api.js?render=6LdyC2cUAAAAACGuDKpXeDorzUDWXmdqeg-xy696" nonce="6Hei0z/rbjQEvwHgCc7wXg=="></script>
```
I appreciate this is a very simple example, so it would be helpful to see what you're putting in your actual policies to understand where this doesn't work.

**Comment by Sora2455** [comment]
@rowan-m, the reason your example works and ours don't is that we're setting style-src and not setting 'unsafe-inline'. The script itself runs, but the styling is broken and we get a bunch of CSP errors.

If you want to know why we're setting style-src, it's because [third-party-css is not safe](https://jakearchibald.com/2018/third-party-css-is-not-safe/).

**Comment by Sora2455** [comment]
@IanOliver Only problem with that is that if Google ever change their CSS, your hashes will no longer match. And Google change their stuff a lot...

**Comment by IanOliver** [comment]
@Sora2455 Yes, this is something I should have mentioned - very true.

As far as I'm aware, the CSS for reCAPTCHA hasn't changed in a long time, and hopefully if v3 is being concentrated on, maybe v2 will be left as-is. Maybe :)

**Comment by arelaxend** [comment]
@IanOliver Hi! I have tried your hashes, it's not working on my app

**Comment by borisschapira** [comment]
@arelaxend some are the same [on my blog](https://github.com/borisschapira/borisschapira.com/blob/master/_data/csp.yml#L86), some aren't. You'd have to test with a Content-Security-Policy-Report-Only to determine yours.

**Comment by arelaxend** [comment]
@borisschapira thank you, i will test this later, since i don't know how to send headers via React

**Comment by Sora2455** [comment]
@arelaxend First, you don't send headers in React. You send them through whatever platform you're hosting React in - Node.JS, ASP MVC, ASP.NET, PHP, Ruby on Rails, etc.

Second, if the hashes are different for different people, Google might very well be delivering different CSS for different people, making hashing near-impossible.

**Comment by arelaxend** [comment]
@sora2455 yes indeed, that’s why I found nothing about this! It uses nodejs. Thank you for the feedback

**Comment by Virgomax** [comment]
@rowan-m I kindly ask: Is there gonna be a solution for this issue? Please, developers don't want to lower their security levels.

**Comment by borisschapira** [comment]
@Sora2455 Oh thanks, I didn't notice.
@Virgomax I don't think there will be a solution, as there is also a bug on Chrome [preventing the hashes from being recognized by the browser](https://bugs.chromium.org/p/chromium/issues/detail?id=546106).

![capture d ecran 2018-09-21 a 09 22 30](https://user-images.githubusercontent.com/284742/45872829-488e6800-bd91-11e8-8a46-52895d722085.png)

**Comment by IanOliver** [comment]
I also raised a bug with Chromium recently. The hashes I'd previously published worked perfectly in Chrome 68, and then were affected by this bug in Chrome 69: https://bugs.chromium.org/p/chromium/issues/detail?id=881847

**Comment by Virgomax** [comment]
@Sora2455   :O   !!! You are right! I just landed here directly from the search engine and I didn't even see the Readme.md.  Looks like I've been wrong my entire life. Thanks for opening my eyes. ;)  I'll go to swear at some other repo.

@borisschapira that's awful. Guard angel, where the hell are you?

**Comment by rowan-m** [comment]
Good news - the service has been updated so `style-src 'unsafe-inline'` is not required. I've tried out a more restrictive CSP on [`csp-example` branch](https://github.com/google/recaptcha/blob/csp-example/examples/recaptcha-v3-request-scores.php#L31-L46) where I'm essentially sending:
```
Content-Security-Policy:
  default-src 'self';
  script-src 'nonce-rand0m1shValue';
  img-src www.gstatic.com;
  frame-src www.google.com;
  object-src 'none';
  base-uri 'none';
```

The `nonce` value is the one being passed to reCAPTCHA in its tag:
```html
<script src="https://www.google.com/recaptcha/api.js?render=your-site-key" nonce="rand0m1shValue"></script>
```

I'll leave this issue open for a little while in case there are any questions / comments and I'll also merge a separate CSP example onto `master`.
